**Clean Flights 2018**
<br/>Perform cleaning on the 2018 dataframe and then export a clean csv.

In [1]:
import numpy as np
import pandas as pd

In [2]:
import warnings
warnings.filterwarnings("ignore")

In [3]:
# increase display size to enable viewing of all data columns

pd.options.display.max_columns = 35

In [4]:
# read in the 2018 flights csv and ensure everything loads correctly

flights2018_df = pd.read_csv('flights2018.csv', index_col=0)
flights2018_df

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
0                5  2018-08-31                UA      N75432    EWR   
1                5  2018-08-31                UA      N461UA    ORD   
2                5  2018-08-31                UA      N76514    SNA   
3                5  2018-08-31                UA      N76533    LAX   
4                5  2018-08-31                UA      N35204    SFO   
...            ...         ...               ...         ...    ...   
7213441          4  2018-06-21                DL      N970AT    ORD   
7213442          4  2018-06-21                DL      N927AT    BNA   
7213443          4  2018-06-21                DL      N927AT    MSP   
7213444          4  2018-06-21                DL      N320NB    AUS   
7213445          4  2018-06-21                DL      N906AT    LGA   

            OriginCityName OriginState Dest        DestCityName DestState  \
0               Newark, NJ          NJ  DEN          Denver, CO        CO   
1              Chicago, IL          IL  GEG         Spokane, WA        WA   
2            Santa Ana, CA          CA  DEN          Denver, CO        CO   
3          Los Angeles, CA          CA  SFO   San Francisco, CA        CA   
4        San Francisco, CA          CA  SAN       San Diego, CA        CA   
...                    ...         ...  ...                 ...       ...   
7213441        Chicago, IL          IL  ATL         Atlanta, GA        GA   
7213442      Nashville, TN          TN  MSP     Minneapolis, MN        MN   
7213443    Minneapolis, MN          MN  BNA       Nashville, TN        TN   
7213444         Austin, TX          TX  SLC  Salt Lake City, UT        UT   
7213445       New York, NY          NY  ORD         Chicago, IL        IL   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
0              1601   1559.0      -2.0       0.0                  -1.0   
1              1932   1926.0      -6.0       0.0                  -1.0   
2              1327   1636.0     189.0       1.0                  12.0   
3               600    550.0     -10.0       0.0                  -1.0   
4              2140   2141.0       1.0       0.0                   0.0   
...             ...      ...       ...       ...                   ...   
7213441        1335   1354.0      19.0       1.0                   1.0   
7213442        1425   1417.0      -8.0       0.0                  -1.0   
7213443        1125   1120.0      -5.0       0.0                  -1.0   
7213444        1905   1931.0      26.0       1.0                   1.0   
7213445        1915   2026.0      71.0       1.0                   4.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
0           17.0     1616.0          0              NaN      1605   
1           45.0     2011.0          0              NaN      1498   
2           20.0     1656.0          0              NaN       846   
3           16.0      606.0          0              NaN       337   
4           31.0     2212.0          0              NaN       447   
...          ...        ...        ...              ...       ...   
7213441     33.0     1427.0          0              NaN       606   
7213442     14.0     1431.0          0              NaN       695   
7213443     14.0     1134.0          0              NaN       695   
7213444      9.0     1940.0          0              NaN      1086   
7213445     37.0     2103.0          0              NaN       733   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0                    7           NaN           NaN       NaN            NaN   
1                    6           NaN           NaN       NaN            NaN   
2                    4           0.0          54.0       5.0            0.0   
3                    2           NaN           NaN       NaN            NaN   
4                    2           NaN           NaN       NaN            NaN   
...                ...           ...   

In [5]:
# summary statistics for the dataframe

flights2018_df.describe().transpose()

count         mean         std    min     25%  \
DayOfWeek             7213446.0     3.934956    1.994790    1.0     2.0   
CRSDepTime            7213446.0  1329.687018  490.931982    1.0   915.0   
DepTime               7101129.0  1333.853806  504.505548    1.0   916.0   
DepDelay              7096212.0     9.969858   44.829641 -122.0    -5.0   
DepDel15              7096212.0     0.184103    0.387568    0.0     0.0   
DepartureDelayGroups  7096212.0     0.040426    2.191893   -2.0    -1.0   
TaxiOut               7097616.0    17.410614    9.920409    1.0    11.0   
WheelsOff             7097617.0  1357.798878  505.972136    1.0   932.0   
Cancelled             7213446.0     0.016162    0.126098    0.0     0.0   
Distance              7213446.0   799.990355  598.178637   31.0   363.0   
DistanceGroup         7213446.0     3.673625    2.358034    1.0     2.0   
CarrierDelay          1352710.0    19.455006   58.908119    0.0     0.0   
WeatherDelay          1352710.0     3.636459   29.996006    0.0     0.0   
NASDelay              1352710.0    15.885471   35.893497    0.0     0.0   
SecurityDelay         1352710.0     0.093539    3.174306    0.0     0.0   
LateAircraftDelay     1352710.0    25.644120   49.787761    0.0     0.0   
CRSArrTime            7213446.0  1486.341099  518.312428    1.0  1100.0   
ArrTime               7094201.0  1466.784165  537.708924    1.0  1049.0   
ArrDelay              7076406.0     5.048581   46.926637 -120.0   -14.0   

                         50%     75%     max  
DayOfWeek                4.0     6.0     7.0  
CRSDepTime            1320.0  1735.0  2359.0  
DepTime               1326.0  1744.0  2400.0  
DepDelay                -2.0     7.0  2710.0  
DepDel15                 0.0     0.0     1.0  
DepartureDelayGroups    -1.0     0.0    12.0  
TaxiOut                 15.0    20.0   196.0  
WheelsOff             1340.0  1759.0  2400.0  
Cancelled                0.0     0.0     1.0  
Distance               632.0  1034.0  4983.0  
DistanceGroup            3.0     5.0    11.0  
CarrierDelay             0.0    17.0  2109.0  
WeatherDelay             0.0     0.0  2692.0  
NASDelay                 3.0    20.0  1848.0  
SecurityDelay            0.0     0.0   987.0  
LateAircraftDelay        3.0    31.0  2454.0  
CRSArrTime            1515.0  1919.0  2400.0  
ArrTime               1506.0  1916.0  2400.0  
ArrDelay                -6.0     8.0  2692.0

In [6]:
# check to see if there any NaN values in the data

flights2018_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number               14526
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                  112317
DepDelay                 117234
DepDel15                 117234
DepartureDelayGroups     117234
TaxiOut                  115830
WheelsOff                115829
Cancelled                     0
CancellationCode        7096862
Distance                      0
DistanceGroup                 0
CarrierDelay            5860736
WeatherDelay            5860736
NASDelay                5860736
SecurityDelay           5860736
LateAircraftDelay       5860736
CRSArrTime                    0
ArrTime                  119245
ArrDelay                 137040
dtype: int64

In [7]:
# list of all the columns in the data dataframe with missing values and the percent that's missing

round(np.mean(flights2018_df[flights2018_df.columns[flights2018_df.isnull().any()]].isna(), axis = 0) * 100,2)

Tail_Number              0.20
DepTime                  1.56
DepDelay                 1.63
DepDel15                 1.63
DepartureDelayGroups     1.63
TaxiOut                  1.61
WheelsOff                1.61
CancellationCode        98.38
CarrierDelay            81.25
WeatherDelay            81.25
NASDelay                81.25
SecurityDelay           81.25
LateAircraftDelay       81.25
ArrTime                  1.65
ArrDelay                 1.90
dtype: float64

In [8]:
# this shows that all flights missing the tail number were cancelled

flights2018_df[(flights2018_df['Cancelled'] == 1) & (flights2018_df['Tail_Number'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
1549             5  2018-08-31                UA         NaN    IND   
1865             5  2018-08-31                UA         NaN    IAD   
3851             3  2018-08-29                UA         NaN    PHL   
4111             3  2018-08-29                UA         NaN    AUS   
4377             3  2018-08-29                UA         NaN    GEG   
...            ...         ...               ...         ...    ...   
7129214          6  2018-06-16                UA         NaN    IAH   
7129252          6  2018-06-16                UA         NaN    MIA   
7129421          6  2018-06-16                UA         NaN    BOS   
7129480          6  2018-06-16                UA         NaN    LGA   
7129666          6  2018-06-16                UA         NaN    IAH   

           OriginCityName OriginState Dest       DestCityName DestState  \
1549     Indianapolis, IN          IN  SFO  San Francisco, CA        CA   
1865       Washington, DC          VA  SFO  San Francisco, CA        CA   
3851     Philadelphia, PA          PA  ORD        Chicago, IL        IL   
4111           Austin, TX          TX  SFO  San Francisco, CA        CA   
4377          Spokane, WA          WA  ORD        Chicago, IL        IL   
...                   ...         ...  ...                ...       ...   
7129214       Houston, TX          TX  SAT    San Antonio, TX        TX   
7129252         Miami, FL          FL  EWR         Newark, NJ        NJ   
7129421        Boston, MA          MA  IAH        Houston, TX        TX   
7129480      New York, NY          NY  ORD        Chicago, IL        IL   
7129666       Houston, TX          TX  BOS         Boston, MA        MA   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
1549            610      NaN       NaN       NaN                   NaN   
1865           2210      NaN       NaN       NaN                   NaN   
3851            700      NaN       NaN       NaN                   NaN   
4111            755      NaN       NaN       NaN                   NaN   
4377            710      NaN       NaN       NaN                   NaN   
...             ...      ...       ...       ...                   ...   
7129214        1155      NaN       NaN       NaN                   NaN   
7129252        1640      NaN       NaN       NaN                   NaN   
7129421        1352      NaN       NaN       NaN                   NaN   
7129480        1100      NaN       NaN       NaN                   NaN   
7129666         801      NaN       NaN       NaN                   NaN   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
1549         NaN        NaN          1                A      1943   
1865         NaN        NaN          1                B      2419   
3851         NaN        NaN          1                B       678   
4111         NaN        NaN          1                A      1504   
4377         NaN        NaN          1                C      1498   
...          ...        ...        ...              ...       ...   
7129214      NaN        NaN          1                A       191   
7129252      NaN        NaN          1                A      1085   
7129421      NaN        NaN          1                A      1597   
7129480      NaN        NaN          1                A       733   
7129666      NaN        NaN          1                A      1597   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1549                 8           NaN           NaN       NaN            NaN   
1865                10           NaN           NaN       NaN            NaN   
3851                 3           NaN           NaN       NaN            NaN   
4111                 7           NaN           NaN       NaN            NaN   
4377                 6           NaN           NaN       NaN            NaN   
...                ...           ...           ...       ...   

In [9]:
# fill in 'unknown' for the NaN tail numbers
# these are all flights that were cancelled, and we already have a small number of cancelled flights, so don't want to delete them

flights2018_df['Tail_Number'].fillna('unknown', inplace=True)
flights2018_df.shape

(7213446, 29)

In [10]:
# check to ensure that tail number was filled in appropriately

flights2018_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                  112317
DepDelay                 117234
DepDel15                 117234
DepartureDelayGroups     117234
TaxiOut                  115830
WheelsOff                115829
Cancelled                     0
CancellationCode        7096862
Distance                      0
DistanceGroup                 0
CarrierDelay            5860736
WeatherDelay            5860736
NASDelay                5860736
SecurityDelay           5860736
LateAircraftDelay       5860736
CRSArrTime                    0
ArrTime                  119245
ArrDelay                 137040
dtype: int64

In [11]:
# check to see how many cancelled flights there were in 2018

flights2018_df[flights2018_df['Cancelled'] == 1].shape

(116584, 29)

116,584 cancelled flights, but 112,317 with no departure time -- there are 4,267 (see below) that need updating.

In [12]:
# look at the flights that are both cancelled and have a departure time

flights2018_df[(flights2018_df['Cancelled'] ==1) & (flights2018_df['DepTime'])]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
6276             2  2018-08-28                UA      N54711    ORD   
6945             5  2018-08-17                UA      N36247    EWR   
7024             5  2018-08-17                UA      N844UA    ORD   
10443            3  2018-08-15                UA      N38417    PDX   
10529            3  2018-08-15                UA      N471UA    ORD   
...            ...         ...               ...         ...    ...   
7164795          6  2018-06-16                DL      N586NW    KOA   
7173743          4  2018-06-28                DL      N955AT    LGA   
7174126          4  2018-06-28                DL      N921DL    ATL   
7175350          4  2018-06-28                DL      N908DA    ATL   
7176189          4  2018-06-28                DL      N904DL    ATL   

        OriginCityName OriginState Dest        DestCityName DestState  \
6276       Chicago, IL          IL  LGA        New York, NY        NY   
6945        Newark, NJ          NJ  ATL         Atlanta, GA        GA   
7024       Chicago, IL          IL  LGA        New York, NY        NY   
10443     Portland, OR          OR  SFO   San Francisco, CA        CA   
10529      Chicago, IL          IL  LGA        New York, NY        NY   
...                ...         ...  ...                 ...       ...   
7164795       Kona, HI          HI  LAX     Los Angeles, CA        CA   
7173743   New York, NY          NY  BOS          Boston, MA        MA   
7174126    Atlanta, GA          GA  PNS       Pensacola, FL        FL   
7175350    Atlanta, GA          GA  SAT     San Antonio, TX        TX   
7176189    Atlanta, GA          GA  RDU  Raleigh/Durham, NC        NC   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
6276           1900   2142.0     162.0       1.0                  10.0   
6945           2136   2315.0      99.0       1.0                   6.0   
7024           1600   2021.0     261.0       1.0                  12.0   
10443           959   1956.0     597.0       1.0                  12.0   
10529           900    856.0      -4.0       0.0                  -1.0   
...             ...      ...       ...       ...                   ...   
7164795        2110   2110.0       NaN       NaN                   NaN   
7173743        1100   1428.0     208.0       1.0                  12.0   
7174126        2034    220.0     346.0       1.0                  12.0   
7175350        1755   2341.0     346.0       1.0                  12.0   
7176189        1354   2111.0     437.0       1.0                  12.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
6276         NaN        NaN          1                B       733   
6945         NaN        NaN          1                B       746   
7024         NaN        NaN          1                B       733   
10443        NaN        NaN          1                A       550   
10529        NaN        NaN          1                A       733   
...          ...        ...        ...              ...       ...   
7164795      NaN        NaN          1                A      2504   
7173743      NaN        NaN          1                A       184   
7174126      NaN        NaN          1                A       271   
7175350      NaN        NaN          1                A       874   
7176189      NaN        NaN          1                A       356   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
6276                 3           NaN           NaN       NaN            NaN   
6945                 3           NaN           NaN       NaN            NaN   
7024                 3           NaN           NaN       NaN            NaN   
10443                3           NaN           NaN       NaN            NaN   
10529                3           NaN           NaN       NaN            NaN   
...                ...           ...           ...       ...            ...   
7164795 

These don't have arrival times, so it is safe to assume that they were actually cancelled, and should not have a departure time.

In [13]:
# change the departure time to 0 for all flights that are cancelled and have a depature time
# and check that there are now no flights with this combination

flights2018_df['DepTime'][(flights2018_df['Cancelled'] == 1) & (flights2018_df['DepTime'])] = 0
flights2018_df[(flights2018_df['Cancelled'] ==1) & (flights2018_df['DepTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [14]:
# check for flights that have a depature time greater than 0, but no departure delay

flights2018_df[(flights2018_df['DepTime'] > 0 ) & (flights2018_df['DepDelay'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
76209            2  2018-08-07                F9      N307FR    CVG   
375523           2  2018-08-21                9E      N319PQ    DSM   
375667           4  2018-08-30                9E      N306PQ    CLT   
375676           4  2018-08-09                9E      N606LR    ATL   
375809           4  2018-08-16                9E      N605LR    LGA   
...            ...         ...               ...         ...    ...   
6609165          6  2018-06-16                9E      N902XJ    MSP   
6609176          4  2018-06-21                9E      N925XJ    ORF   
6609193          7  2018-06-17                9E      N919XJ    LGA   
7066206          2  2018-06-05                F9      N943FR    GRR   
7067252          6  2018-06-09                F9      N304FR    MCO   

           OriginCityName OriginState Dest           DestCityName DestState  \
76209      Cincinnati, OH          KY  LGA           New York, NY        NY   
375523     Des Moines, IA          IA  MSP        Minneapolis, MN        MN   
375667      Charlotte, NC          NC  LGA           New York, NY        NY   
375676        Atlanta, GA          GA  TLH        Tallahassee, FL        FL   
375809       New York, NY          NY  BHM         Birmingham, AL        AL   
...                   ...         ...  ...                    ...       ...   
6609165   Minneapolis, MN          MN  DFW  Dallas/Fort Worth, TX        TX   
6609176       Norfolk, VA          VA  JFK           New York, NY        NY   
6609193      New York, NY          NY  BHM         Birmingham, AL        AL   
7066206  Grand Rapids, MI          MI  DEN             Denver, CO        CO   
7067252       Orlando, FL          FL  OKC      Oklahoma City, OK        OK   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
76209          2002   2032.0       NaN       NaN                   NaN   
375523         1109   1109.0       NaN       NaN                   NaN   
375667         2001   2001.0       NaN       NaN                   NaN   
375676         2240   2240.0       NaN       NaN                   NaN   
375809         2045   2045.0       NaN       NaN                   NaN   
...             ...      ...       ...       ...                   ...   
6609165        1115   1115.0       NaN       NaN                   NaN   
6609176        1022   1022.0       NaN       NaN                   NaN   
6609193        1422   1422.0       NaN       NaN                   NaN   
7066206        1900   1951.0       NaN       NaN                   NaN   
7067252        1850   1858.0       NaN       NaN                   NaN   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
76209       21.0     2053.0          0              NaN       585   
375523      51.0     1200.0          0              NaN       232   
375667      33.0     2034.0          0              NaN       544   
375676      31.0     2311.0          0              NaN       223   
375809      16.0     2101.0          0              NaN       866   
...          ...        ...        ...              ...       ...   
6609165     22.0     1137.0          0              NaN       852   
6609176     17.0     1039.0          0              NaN       290   
6609193     36.0     1458.0          0              NaN       866   
7066206     30.0     2021.0          0              NaN      1015   
7067252     10.0     1908.0          0              NaN      1069   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
76209                3           NaN           NaN       NaN            NaN   
375523               1           0.0           0.0      30.0            0.0   
375667               3           NaN           NaN       NaN            NaN   
375676               1           NaN           NaN       NaN            NaN   
375809               4           NaN           NaN       NaN            NaN   
...            

In [15]:
# fill in the appropriate values for DepDelay, DepDel15, and DepatureDelayGroups

# DepDelay = DepTime - CRSDepTime
flights2018_df['DepDelay'][(flights2018_df['DepTime'] > 0 ) & (flights2018_df['DepDelay'].isna())] = flights2018_df['DepTime'] - flights2018_df['CRSDepTime']

# DepDel15 0 or 1 depending on whether the delay is less or more than 15 minutes
flights2018_df.loc[flights2018_df['DepDelay'] <= 15 & (flights2018_df['DepDel15'].isna()), 'DepDel15'] = 0
flights2018_df.loc[flights2018_df['DepDelay'] > 15 & (flights2018_df['DepDel15'].isna()), 'DepDel15'] = 1

# DepartureDelayGroups based on 15 minute intervals
flights2018_df.loc[(flights2018_df['DepDelay'] < -59) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 0
flights2018_df.loc[(flights2018_df['DepDelay'] >= -15) & (flights2018_df['DepDelay'] < 0) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 0
flights2018_df.loc[(flights2018_df['DepDelay'] >= 0) & (flights2018_df['DepDelay'] < 15) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 0
flights2018_df.loc[(flights2018_df['DepDelay'] >= 15) & (flights2018_df['DepDelay'] < 30) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 1
flights2018_df.loc[(flights2018_df['DepDelay'] >= 30) & (flights2018_df['DepDelay'] < 45)  & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 2
flights2018_df.loc[(flights2018_df['DepDelay'] >= 45) & (flights2018_df['DepDelay'] < 60) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 3
flights2018_df.loc[(flights2018_df['DepDelay'] >= 60) & (flights2018_df['DepDelay'] < 75) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 4
flights2018_df.loc[(flights2018_df['DepDelay'] >= 75) & (flights2018_df['DepDelay'] < 90) &  (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 5
flights2018_df.loc[(flights2018_df['DepDelay'] >= 90) & (flights2018_df['DepDelay'] < 105) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 6
flights2018_df.loc[(flights2018_df['DepDelay'] >= 105) & (flights2018_df['DepDelay'] < 120) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 7
flights2018_df.loc[(flights2018_df['DepDelay'] >= 120) & (flights2018_df['DepDelay'] < 135) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 8
flights2018_df.loc[(flights2018_df['DepDelay'] >= 135) & (flights2018_df['DepDelay'] < 150) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 9
flights2018_df.loc[(flights2018_df['DepDelay'] >= 150) & (flights2018_df['DepDelay'] < 165) & (flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 10
flights2018_df.loc[(flights2018_df['DepDelay'] >= 165) & (flights2018_df['DepDelay'] < 180) &(flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 11
flights2018_df.loc[(flights2018_df['DepDelay'] >= 180) &(flights2018_df['DepartureDelayGroups'].isna()), 'DepartureDelayGroups'] = 12

flights2018_df[(flights2018_df['DepTime'] > 0 ) & (flights2018_df['DepartureDelayGroups'].isna())]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [16]:
# review remaining flights with NaN values in the DepDelay column

flights2018_df[(flights2018_df['DepDelay'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
1549             5  2018-08-31                UA     unknown    IND   
1865             5  2018-08-31                UA     unknown    IAD   
3851             3  2018-08-29                UA     unknown    PHL   
4111             3  2018-08-29                UA     unknown    AUS   
4377             3  2018-08-29                UA     unknown    GEG   
...            ...         ...               ...         ...    ...   
7194535          5  2018-06-29                DL      N932DL    CLE   
7194647          5  2018-06-29                DL      N935DL    EWR   
7194920          5  2018-06-29                DL      N951DN    CHS   
7207659          2  2018-06-19                DL      N980AT    LGA   
7207677          2  2018-06-19                DL      N980AT    ORD   

           OriginCityName OriginState Dest       DestCityName DestState  \
1549     Indianapolis, IN          IN  SFO  San Francisco, CA        CA   
1865       Washington, DC          VA  SFO  San Francisco, CA        CA   
3851     Philadelphia, PA          PA  ORD        Chicago, IL        IL   
4111           Austin, TX          TX  SFO  San Francisco, CA        CA   
4377          Spokane, WA          WA  ORD        Chicago, IL        IL   
...                   ...         ...  ...                ...       ...   
7194535     Cleveland, OH          OH  ATL        Atlanta, GA        GA   
7194647        Newark, NJ          NJ  ATL        Atlanta, GA        GA   
7194920    Charleston, SC          SC  ATL        Atlanta, GA        GA   
7207659      New York, NY          NY  ORD        Chicago, IL        IL   
7207677       Chicago, IL          IL  LGA       New York, NY        NY   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
1549            610      NaN       NaN       NaN                   NaN   
1865           2210      NaN       NaN       NaN                   NaN   
3851            700      NaN       NaN       NaN                   NaN   
4111            755      NaN       NaN       NaN                   NaN   
4377            710      NaN       NaN       NaN                   NaN   
...             ...      ...       ...       ...                   ...   
7194535         700      NaN       NaN       NaN                   NaN   
7194647         600      NaN       NaN       NaN                   NaN   
7194920         545      NaN       NaN       NaN                   NaN   
7207659        1715      NaN       NaN       NaN                   NaN   
7207677        1930      NaN       NaN       NaN                   NaN   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
1549         NaN        NaN          1                A      1943   
1865         NaN        NaN          1                B      2419   
3851         NaN        NaN          1                B       678   
4111         NaN        NaN          1                A      1504   
4377         NaN        NaN          1                C      1498   
...          ...        ...        ...              ...       ...   
7194535      NaN        NaN          1                A       554   
7194647      NaN        NaN          1                A       746   
7194920      NaN        NaN          1                A       259   
7207659      NaN        NaN          1                C       733   
7207677      NaN        NaN          1                C       733   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1549                 8           NaN           NaN       NaN            NaN   
1865                10           NaN           NaN       NaN            NaN   
3851                 3           NaN           NaN       NaN            NaN   
4111                 7           NaN           NaN       NaN            NaN   
4377                 6           NaN           NaN       NaN            NaN   
...                ...           ...           ...       ...   

In [17]:
# fill in 0 for the remaining NaN values in the DepTime, DepDelay, DepDel15 and DepartureDelayGroups columns

flights2018_df['DepTime'].fillna(0, inplace=True)
flights2018_df['DepDelay'].fillna(0, inplace=True)
flights2018_df['DepDel15'].fillna(0, inplace=True)
flights2018_df['DepartureDelayGroups'].fillna(0, inplace=True)
flights2018_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                  115830
WheelsOff                115829
Cancelled                     0
CancellationCode        7096862
Distance                      0
DistanceGroup                 0
CarrierDelay            5860736
WeatherDelay            5860736
NASDelay                5860736
SecurityDelay           5860736
LateAircraftDelay       5860736
CRSArrTime                    0
ArrTime                  119245
ArrDelay                 137040
dtype: int64

In [18]:
# look at the flights that are both cancelled and have a arrival time

flights2018_df[(flights2018_df['Cancelled'] == 1) & (flights2018_df['ArrTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [19]:
# inspect the columns with NaN values for arrival time that are not cancelled

flights2018_df[(flights2018_df['Cancelled'] == 0) & (flights2018_df['ArrTime'].isna())]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
5945             2  2018-08-28                UA      N76516    IAH   
6035             2  2018-08-28                UA      N848UA    BDL   
6101             2  2018-08-28                UA      N451UA    JAC   
8546             4  2018-08-16                UA      N15710    EWR   
9785             3  2018-08-15                UA      N417UA    ORD   
...            ...         ...               ...         ...    ...   
7151798          2  2018-06-26                DL      N957AT    ORD   
7170071          2  2018-06-26                DL      N956AT    ORD   
7175052          4  2018-06-28                DL      N990DL    MIA   
7185018          2  2018-06-12                DL      N357NB    DFW   
7206521          1  2018-06-18                DL      N718TW    SEA   

                OriginCityName OriginState Dest   DestCityName DestState  \
5945               Houston, TX          TX  ORD    Chicago, IL        IL   
6035              Hartford, CT          CT  ORD    Chicago, IL        IL   
6101               Jackson, WY          WY  ORD    Chicago, IL        IL   
8546                Newark, NJ          NJ  SNA  Santa Ana, CA        CA   
9785               Chicago, IL          IL  SNA  Santa Ana, CA        CA   
...                        ...         ...  ...            ...       ...   
7151798            Chicago, IL          IL  LGA   New York, NY        NY   
7170071            Chicago, IL          IL  LGA   New York, NY        NY   
7175052              Miami, FL          FL  ATL    Atlanta, GA        GA   
7185018  Dallas/Fort Worth, TX          TX  LGA   New York, NY        NY   
7206521            Seattle, WA          WA  JFK   New York, NY        NY   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
5945           1819   1904.0      45.0       1.0                   3.0   
6035           1820   2149.0     209.0       1.0                  12.0   
6101           1256   1714.0     258.0       1.0                  12.0   
8546           1835   2159.0     204.0       1.0                  12.0   
9785           1945   2227.0     162.0       1.0                  10.0   
...             ...      ...       ...       ...                   ...   
7151798        1930     26.0     296.0       1.0                  12.0   
7170071        1530   2214.0     404.0       1.0                  12.0   
7175052        1110   1112.0       2.0       1.0                   0.0   
7185018        1900   2211.0     191.0       1.0                  12.0   
7206521         740    740.0       0.0       0.0                   0.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
5945        14.0     1918.0          0              NaN       925   
6035        11.0     2200.0          0              NaN       783   
6101        16.0     1730.0          0              NaN      1162   
8546        17.0     2216.0          0              NaN      2434   
9785        28.0     2255.0          0              NaN      1726   
...          ...        ...        ...              ...       ...   
7151798     23.0       49.0          0              NaN       733   
7170071     31.0     2245.0          0              NaN       733   
7175052     17.0     1129.0          0              NaN       594   
7185018     12.0     2223.0          0              NaN      1389   
7206521     25.0      805.0          0              NaN      2422   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
5945                 4           NaN           NaN       NaN            NaN   
6035                 4           NaN           NaN       NaN            NaN   
6101                 5           NaN           NaN       NaN            NaN   
8546                10           NaN           NaN       NaN            NaN   
9785                 7           NaN           NaN       NaN            NaN   
...                ...           ...           ... 

In [20]:
# there are 10,318 rows with flights that are not cancelled, and have a departure time, but no arrival time
# delete these rows

flights2018_df = flights2018_df.drop(flights2018_df[(flights2018_df['Cancelled'] == 0) & (flights2018_df['ArrTime'].isna())].index)
flights2018_df[(flights2018_df['Cancelled'] == 0) & (flights2018_df['ArrTime'].isna())]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [21]:
# replace NaN with 0 for all flights that are cancelled 
# and check that there are now no flights with this combination

flights2018_df['ArrTime'][(flights2018_df['Cancelled'] == 1) & (flights2018_df['ArrTime'].isna())] = 0
flights2018_df['ArrDelay'][(flights2018_df['Cancelled'] == 1) & (flights2018_df['ArrDelay'].isna())] = 0
flights2018_df[(flights2018_df['Cancelled'] == 1) & (flights2018_df['ArrTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [22]:
# check the status of NaNs

flights2018_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                  115830
WheelsOff                115829
Cancelled                     0
CancellationCode        7094201
Distance                      0
DistanceGroup                 0
CarrierDelay            5858075
WeatherDelay            5858075
NASDelay                5858075
SecurityDelay           5858075
LateAircraftDelay       5858075
CRSArrTime                    0
ArrTime                       0
ArrDelay                  17795
dtype: int64

In [23]:
# look at the remaining flights with NaN values for ArrDelay

flights2018_df[(flights2018_df['ArrDelay'].isna()) & (flights2018_df['ArrTime'])]

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
110              5  2018-08-31                UA      N69816    ATL   
114              5  2018-08-31                UA      N434UA    SLC   
184              5  2018-08-31                UA      N77430    PDX   
515              5  2018-08-31                UA      N411UA    EWR   
568              5  2018-08-31                UA      N495UA    DEN   
...            ...         ...               ...         ...    ...   
7208961          2  2018-06-19                DL      N333DX    ATL   
7208985          2  2018-06-19                DL      N932AT    SEA   
7210074          2  2018-06-19                DL      N938AT    ORD   
7211514          3  2018-06-20                DL      N958AT    MSP   
7212227          3  2018-06-20                DL      N916DL    TPA   

             OriginCityName OriginState Dest           DestCityName DestState  \
110             Atlanta, GA          GA  DEN             Denver, CO        CO   
114      Salt Lake City, UT          UT  DEN             Denver, CO        CO   
184            Portland, OR          OR  DEN             Denver, CO        CO   
515              Newark, NJ          NJ  TPA              Tampa, FL        FL   
568              Denver, CO          CO  OMA              Omaha, NE        NE   
...                     ...         ...  ...                    ...       ...   
7208961         Atlanta, GA          GA  BOS             Boston, MA        MA   
7208985         Seattle, WA          WA  DEN             Denver, CO        CO   
7210074         Chicago, IL          IL  LGA           New York, NY        NY   
7211514     Minneapolis, MN          MN  DFW  Dallas/Fort Worth, TX        TX   
7212227           Tampa, FL          FL  ATL            Atlanta, GA        GA   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
110            1641   1637.0      -4.0       0.0                  -1.0   
114            1534   1533.0      -1.0       0.0                  -1.0   
184            1330   1334.0       4.0       1.0                   0.0   
515            1340   1339.0      -1.0       0.0                  -1.0   
568            1753   2226.0     273.0       1.0                  12.0   
...             ...      ...       ...       ...                   ...   
7208961        1610   1615.0       5.0       1.0                   0.0   
7208985        1525   1533.0       8.0       1.0                   0.0   
7210074        1530   2312.0     462.0       1.0                  12.0   
7211514        1930   1928.0      -2.0       0.0                  -1.0   
7212227        1315   1356.0      41.0       1.0                   2.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
110         18.0     1655.0          0              NaN      1199   
114         14.0     1547.0          0              NaN       391   
184         13.0     1347.0          0              NaN       991   
515         69.0     1448.0          0              NaN       997   
568         13.0     2239.0          0              NaN       472   
...          ...        ...        ...              ...       ...   
7208961     15.0     1630.0          0              NaN       946   
7208985     27.0     1600.0          0              NaN      1024   
7210074     11.0     2323.0          0              NaN       733   
7211514     14.0     1942.0          0              NaN       852   
7212227      9.0     1405.0          0              NaN       406   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
110                  5           NaN           NaN       NaN            NaN   
114                  2           NaN           NaN       NaN            NaN   
184                  4           NaN           NaN       NaN            NaN   
515                  4           NaN           NaN       NaN            NaN   
568                  2           NaN           NaN       NaN          

In [24]:
# fill in these NaN values with the difference between the expected arrival time (CRSArrTime) and the actual arrival time (ArrTime)

flights2018_df['ArrDelay'].fillna((flights2018_df['ArrTime'] - flights2018_df['CRSArrTime']), inplace=True)
flights2018_df[(flights2018_df['ArrDelay'].isna()) & (flights2018_df['ArrTime'])]

Empty DataFrame
Columns: [DayOfWeek, FlightDate, Reporting_Airline, Tail_Number, Origin, OriginCityName, OriginState, Dest, DestCityName, DestState, CRSDepTime, DepTime, DepDelay, DepDel15, DepartureDelayGroups, TaxiOut, WheelsOff, Cancelled, CancellationCode, Distance, DistanceGroup, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay, CRSArrTime, ArrTime, ArrDelay]
Index: []

In [25]:
# check the status of NaNs

flights2018_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                  115830
WheelsOff                115829
Cancelled                     0
CancellationCode        7094201
Distance                      0
DistanceGroup                 0
CarrierDelay            5858075
WeatherDelay            5858075
NASDelay                5858075
SecurityDelay           5858075
LateAircraftDelay       5858075
CRSArrTime                    0
ArrTime                       0
ArrDelay                      0
dtype: int64

In [26]:
# fill in the TaxiOut column with the median value 

flights2018_df['TaxiOut'].fillna(flights2018_df['TaxiOut'].median(), inplace=True)
flights2018_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                       0
WheelsOff                115829
Cancelled                     0
CancellationCode        7094201
Distance                      0
DistanceGroup                 0
CarrierDelay            5858075
WeatherDelay            5858075
NASDelay                5858075
SecurityDelay           5858075
LateAircraftDelay       5858075
CRSArrTime                    0
ArrTime                       0
ArrDelay                      0
dtype: int64

In [27]:
# fill in WheelsOff with the CRSDepTime + the TaxiOut time

flights2018_df['WheelsOff'].fillna((flights2018_df['CRSDepTime'] + flights2018_df['TaxiOut']), inplace=True)
flights2018_df.isna().sum()

DayOfWeek                     0
FlightDate                    0
Reporting_Airline             0
Tail_Number                   0
Origin                        0
OriginCityName                0
OriginState                   0
Dest                          0
DestCityName                  0
DestState                     0
CRSDepTime                    0
DepTime                       0
DepDelay                      0
DepDel15                      0
DepartureDelayGroups          0
TaxiOut                       0
WheelsOff                     0
Cancelled                     0
CancellationCode        7094201
Distance                      0
DistanceGroup                 0
CarrierDelay            5858075
WeatherDelay            5858075
NASDelay                5858075
SecurityDelay           5858075
LateAircraftDelay       5858075
CRSArrTime                    0
ArrTime                       0
ArrDelay                      0
dtype: int64

In [28]:
# check the shapes of the whole dataframe and just the cancelled flights

display(flights2018_df.shape)
display(flights2018_df[flights2018_df['Cancelled'] == 1].shape)

(7210785, 29)

(116584, 29)

116,584 cancelled + 7,094,201 NaN Cancelation Code = 7,201,785 total

In [29]:
# replace NaN with 0 for CancellationCode and the reason for Delay

flights2018_df['CancellationCode'].fillna(0, inplace=True)
flights2018_df['CarrierDelay'].fillna(0, inplace=True)
flights2018_df['WeatherDelay'].fillna(0, inplace=True)
flights2018_df['NASDelay'].fillna(0, inplace=True)
flights2018_df['SecurityDelay'].fillna(0, inplace=True)
flights2018_df['LateAircraftDelay'].fillna(0, inplace=True)
flights2018_df.isna().sum()

DayOfWeek               0
FlightDate              0
Reporting_Airline       0
Tail_Number             0
Origin                  0
OriginCityName          0
OriginState             0
Dest                    0
DestCityName            0
DestState               0
CRSDepTime              0
DepTime                 0
DepDelay                0
DepDel15                0
DepartureDelayGroups    0
TaxiOut                 0
WheelsOff               0
Cancelled               0
CancellationCode        0
Distance                0
DistanceGroup           0
CarrierDelay            0
WeatherDelay            0
NASDelay                0
SecurityDelay           0
LateAircraftDelay       0
CRSArrTime              0
ArrTime                 0
ArrDelay                0
dtype: int64

In [30]:
# add a boolean column for whether a flight is delayed or not (based on whether it arrives late)

flights2018_df['Delayed'] = 0

flights2018_df.loc[flights2018_df['ArrDelay'] <= 0, 'Delayed'] = 0
flights2018_df.loc[flights2018_df['ArrDelay'] > 0, 'Delayed'] = 1

flights2018_df['Delayed'].value_counts()

0    4682585
1    2528200
Name: Delayed, dtype: int64

In [31]:
# display the clean dataframe

flights2018_df

DayOfWeek  FlightDate Reporting_Airline Tail_Number Origin  \
0                5  2018-08-31                UA      N75432    EWR   
1                5  2018-08-31                UA      N461UA    ORD   
2                5  2018-08-31                UA      N76514    SNA   
3                5  2018-08-31                UA      N76533    LAX   
4                5  2018-08-31                UA      N35204    SFO   
...            ...         ...               ...         ...    ...   
7213441          4  2018-06-21                DL      N970AT    ORD   
7213442          4  2018-06-21                DL      N927AT    BNA   
7213443          4  2018-06-21                DL      N927AT    MSP   
7213444          4  2018-06-21                DL      N320NB    AUS   
7213445          4  2018-06-21                DL      N906AT    LGA   

            OriginCityName OriginState Dest        DestCityName DestState  \
0               Newark, NJ          NJ  DEN          Denver, CO        CO   
1              Chicago, IL          IL  GEG         Spokane, WA        WA   
2            Santa Ana, CA          CA  DEN          Denver, CO        CO   
3          Los Angeles, CA          CA  SFO   San Francisco, CA        CA   
4        San Francisco, CA          CA  SAN       San Diego, CA        CA   
...                    ...         ...  ...                 ...       ...   
7213441        Chicago, IL          IL  ATL         Atlanta, GA        GA   
7213442      Nashville, TN          TN  MSP     Minneapolis, MN        MN   
7213443    Minneapolis, MN          MN  BNA       Nashville, TN        TN   
7213444         Austin, TX          TX  SLC  Salt Lake City, UT        UT   
7213445       New York, NY          NY  ORD         Chicago, IL        IL   

         CRSDepTime  DepTime  DepDelay  DepDel15  DepartureDelayGroups  \
0              1601   1559.0      -2.0       0.0                  -1.0   
1              1932   1926.0      -6.0       0.0                  -1.0   
2              1327   1636.0     189.0       1.0                  12.0   
3               600    550.0     -10.0       0.0                  -1.0   
4              2140   2141.0       1.0       1.0                   0.0   
...             ...      ...       ...       ...                   ...   
7213441        1335   1354.0      19.0       1.0                   1.0   
7213442        1425   1417.0      -8.0       0.0                  -1.0   
7213443        1125   1120.0      -5.0       0.0                  -1.0   
7213444        1905   1931.0      26.0       1.0                   1.0   
7213445        1915   2026.0      71.0       1.0                   4.0   

         TaxiOut  WheelsOff  Cancelled CancellationCode  Distance  \
0           17.0     1616.0          0                0      1605   
1           45.0     2011.0          0                0      1498   
2           20.0     1656.0          0                0       846   
3           16.0      606.0          0                0       337   
4           31.0     2212.0          0                0       447   
...          ...        ...        ...              ...       ...   
7213441     33.0     1427.0          0                0       606   
7213442     14.0     1431.0          0                0       695   
7213443     14.0     1134.0          0                0       695   
7213444      9.0     1940.0          0                0      1086   
7213445     37.0     2103.0          0                0       733   

         DistanceGroup  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0                    7           0.0           0.0       0.0            0.0   
1                    6           0.0           0.0       0.0            0.0   
2                    4           0.0          54.0       5.0            0.0   
3                    2           0.0           0.0       0.0            0.0   
4                    2           0.0           0.0       0.0            0.0   
...                ...           ...   

In [32]:
# export the clean dataframe with no NaN values to a csv

flights2018_df.to_csv('flights2018clean.csv')